From 92a39c097559dcd73c510fc3a89405d28cfc9951 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Matthew=20=F0=9F=AA=B7?= <matthew@urgero.net>
Date: Sat, 14 Dec 2024 11:33:42 -0600
Subject: [PATCH 196/196] Add activation code for Judges upgrade (#314)

Make all necessary preparations for the Judges upgrade June 20 2024.
Used previous Joshua upgrade commits as template for this commit.

---------
Co-authored-by: Shammah Chancellor <213740+schancel@users.noreply.github.com>
---
 src/chainparams.cpp                           |  8 ++++
 src/consensus/activation.cpp                  | 10 +++++
 src/consensus/activation.h                    |  4 ++
 src/consensus/addresses.h                     |  4 ++
 src/consensus/params.h                        |  4 ++
 src/init.cpp                                  |  2 +-
 src/minerfund.cpp                             |  5 +++
 src/validation.cpp                            |  2 +-
 .../logos_feature_minerfund_activation.py     | 42 ++++++++++++++++++-
 9 files changed, 78 insertions(+), 3 deletions(-)

diff --git a/src/chainparams.cpp b/src/chainparams.cpp
index ebcbdca29..60ceb6d5b 100644
--- a/src/chainparams.cpp
+++ b/src/chainparams.cpp
@@ -115,6 +115,8 @@ public:
         consensus.joshuaActivationTime = 1703215620;
         // 2024-06-20T20:51:00.000Z protocol upgrade
         consensus.judgesActivationTime = 1718916660;
+        // 2024-12-21T09:20:00.000Z protocol upgrade
+        consensus.ruthActivationTime = 1734772800;
 
         /**
          * The message start string is designed to be unlikely to occur in
@@ -262,6 +264,10 @@ public:
         consensus.judgesActivationTime =
             mainnetConsensus.judgesActivationTime -
             testnetActivationOffset;
+        // 2024-12-21T09:20:00.000Z protocol upgrade
+        consensus.ruthActivationTime =
+            mainnetConsensus.ruthActivationTime -
+            testnetActivationOffset;
 
         // "ltdk" with MSB set
         diskMagic[0] = 0xec;
@@ -376,6 +382,8 @@ public:
         consensus.joshuaActivationTime = mainnetConsensus.joshuaActivationTime;
         // 2024-06-20T20:51:00.000Z protocol upgrade
         consensus.judgesActivationTime = mainnetConsensus.judgesActivationTime;
+        // 2024-12-21T09:20:00.000Z protocol upgrade
+        consensus.ruthActivationTime = mainnetConsensus.ruthActivationTime;
 
         // "lrdk" with MSB set
         diskMagic[0] = 0xec;
diff --git a/src/consensus/activation.cpp b/src/consensus/activation.cpp
index 9b0a4b777..4154e3729 100644
--- a/src/consensus/activation.cpp
+++ b/src/consensus/activation.cpp
@@ -62,3 +62,13 @@ bool IsJoshuaEnabled(const Consensus::Params &params,
     return pindexPrev->GetMedianTimePast() >=
            gArgs.GetArg("-joshuaactivationtime", params.joshuaActivationTime);
 }
+
+bool IsJudgesEnabled(const Consensus::Params &params,
+                          const CBlockIndex *pindexPrev) {
+    if (pindexPrev == nullptr) {
+        return false;
+    }
+
+    return pindexPrev->GetMedianTimePast() >=
+           gArgs.GetArg("-judgesactivationtime", params.judgesActivationTime);
+}
diff --git a/src/consensus/activation.h b/src/consensus/activation.h
index 16e2eee2b..2512ec813 100644
--- a/src/consensus/activation.h
+++ b/src/consensus/activation.h
@@ -35,4 +35,8 @@ bool IsDeuteronomyEnabled(const Consensus::Params &params,
 bool IsJoshuaEnabled(const Consensus::Params &params,
                           const CBlockIndex *pindexPrev);
 
+/** Check if June 20th, 2024 protocol upgrade has activated. */
+bool IsJudgesEnabled(const Consensus::Params &params,
+                          const CBlockIndex *pindexPrev);
+
 #endif // BITCOIN_CONSENSUS_ACTIVATION_H
diff --git a/src/consensus/addresses.h b/src/consensus/addresses.h
index a971ed6f7..3de817010 100644
--- a/src/consensus/addresses.h
+++ b/src/consensus/addresses.h
@@ -137,5 +137,9 @@ Consensus::CoinbaseAddresses AddressSets = {
         {
             "lotus_16PSJJYp3YkVyctW8LRbp6UhbsrQMJghL3jjze3b7",
         },
+    .judges =
+        {
+            "lotus_16PSJMaps9sQg7aBQgyY1RdHb2fZYdmWhQPbgus75"
+        }
 };
 } // namespace RewardAddresses
diff --git a/src/consensus/params.h b/src/consensus/params.h
index 9a6294b66..42062d954 100644
--- a/src/consensus/params.h
+++ b/src/consensus/params.h
@@ -21,6 +21,7 @@ struct CoinbaseAddresses {
     std::vector<std::string> numbers;
     std::vector<std::string> deuteronomy;
     std::vector<std::string> joshua;
+    std::vector<std::string> judges;
 };
 
 /**
@@ -47,6 +48,9 @@ struct Params {
     /** Unix time used for MTP activation of 2024-06-20T20:51:00.000Z protocol
      * upgrade */
     int judgesActivationTime;
+    /** Unit time used for MTP activation of 2024-12-21T09:20:00.000Z protocol
+     * upgrade */
+    int ruthActivationTime;
 
     /**
      * Don't warn about unknown BIP 9 activations below this height.
diff --git a/src/init.cpp b/src/init.cpp
index ef3a6e912..1e048b2e5 100644
--- a/src/init.cpp
+++ b/src/init.cpp
@@ -426,7 +426,7 @@ void SetupServerArgs(NodeContext &node) {
         // TODO remove after the December 2021 upgrade
         "-exodusactivationtime", "-leviticusactivationtime",
         "-numbersactivationtime", "-deuteronomyactivationtime",
-        "-joshuaactivationtime"};
+        "-joshuaactivationtime", "-judgesactivationtime"};
 
     // Set all of the args and their help
     // When adding new options to the categories, please keep and ensure
diff --git a/src/minerfund.cpp b/src/minerfund.cpp
index 25e3c83c9..288294a2f 100644
--- a/src/minerfund.cpp
+++ b/src/minerfund.cpp
@@ -56,6 +56,11 @@ std::vector<CTxOut> GetMinerFundRequiredOutputs(const Consensus::Params &params,
         return {};
     }
 
+    if (IsJudgesEnabled(params, pindexPrev)) {
+        return BuildOutputsCycling(params.coinbasePayoutAddresses.judges,
+                                   pindexPrev, blockReward);
+    }
+
     if (IsJoshuaEnabled(params, pindexPrev)) {
         return BuildOutputsCycling(params.coinbasePayoutAddresses.joshua,
                                    pindexPrev, blockReward);
diff --git a/src/validation.cpp b/src/validation.cpp
index 8b71d13c1..a7d6aca93 100644
--- a/src/validation.cpp
+++ b/src/validation.cpp
@@ -278,7 +278,7 @@ bool CheckSequenceLocks(const CTxMemPool &pool, const CTransaction &tx,
 static bool IsReplayProtectionEnabled(const Consensus::Params &params,
                                       int64_t nMedianTimePast) {
     return nMedianTimePast >= gArgs.GetArg("-replayprotectionactivationtime",
-                                           params.judgesActivationTime);
+                                           params.ruthActivationTime);
 }
 
 static bool IsReplayProtectionEnabled(const Consensus::Params &params,
diff --git a/test/functional/logos_feature_minerfund_activation.py b/test/functional/logos_feature_minerfund_activation.py
index 514237166..7860afd01 100755
--- a/test/functional/logos_feature_minerfund_activation.py
+++ b/test/functional/logos_feature_minerfund_activation.py
@@ -31,8 +31,9 @@ NUMBERS_ACTIVATION_TIME = 2020000000
 DEUTERONOMY_ACTIVATION_TIME = 2030000000
 JOSHUA_ACTIVATION_TIME = 2040000000
 JUDGES_ACTIVATION_TIME = 2050000000
+RUTH_ACTIVATION_TIME = 2060000000
 
-REPLAYPROTECTION_ACTIVATION_TIME = JUDGES_ACTIVATION_TIME
+REPLAYPROTECTION_ACTIVATION_TIME = RUTH_ACTIVATION_TIME
 
 # see consensus/addresses.h, use getaddressinfo to get the scriptPubKey
 GENESIS_SCRIPTS = [
@@ -122,6 +123,9 @@ DEUTERONOMY_SCRIPTS = [
 JOSHUA_SCRIPTS = [
     "76a9142eedc5ad3d142181417daea01076846edfba998088ac",
 ]
+JUDGES_SCRIPTS = [
+    "76a9149208ecc785c92968481a92aee7024b77e54d27dc88ac",
+]
 
 class MinerFundActivationTest(BitcoinTestFramework):
     def set_test_params(self):
@@ -134,6 +138,7 @@ class MinerFundActivationTest(BitcoinTestFramework):
             f'-numbersactivationtime={NUMBERS_ACTIVATION_TIME}',
             f'-deuteronomyactivationtime={DEUTERONOMY_ACTIVATION_TIME}',
             f'-joshuaactivationtime={JOSHUA_ACTIVATION_TIME}',
+            f'-judgesactivationtime={JUDGES_ACTIVATION_TIME}',
             f'-replayprotectionactivationtime={REPLAYPROTECTION_ACTIVATION_TIME}',
         ]]
 
@@ -348,6 +353,41 @@ class MinerFundActivationTest(BitcoinTestFramework):
             prepare_block(block)
             assert_equal(node.submitblock(ToHex(block)), None)
 
+        # Using the judges addresses before upgrade fails
+        block = make_block_cb_post_numbers(JUDGES_SCRIPTS)
+        prepare_block(block)
+        assert_equal(node.submitblock(ToHex(block)), 'bad-cb-minerfund')
+        node.setmocktime(JUDGES_ACTIVATION_TIME)
+
+        # Mine 11 blocks with JUDGES_ACTIVATION in the middle
+        # That moves MTP exactly to JUDGES_ACTIVATION
+        for i in range(-6, 6):
+            block = make_block_cb_post_numbers(JOSHUA_SCRIPTS)
+            block.nTime = JUDGES_ACTIVATION_TIME + i
+            prepare_block(block)
+            assert_equal(node.submitblock(ToHex(block)), None)
+
+        assert_equal(node.getblockchaininfo()['mediantime'],
+                     JUDGES_ACTIVATION_TIME)
+        
+        # Now the using the genesis, exodus, leviticus, or numbers, and deuteronomy addresses fails
+        for block in [
+            make_block_with_cb_scripts(GENESIS_SCRIPTS),
+            make_block_with_cb_scripts(EXODUS_SCRIPTS),
+            make_block_with_cb_scripts(LEVITICUS_SCRIPTS),
+            make_block_cb_post_numbers(NUMBERS_SCRIPTS),
+            make_block_cb_post_numbers(DEUTERONOMY_SCRIPTS),
+            make_block_cb_post_numbers(JOSHUA_SCRIPTS),
+        ]:
+            prepare_block(block)
+            assert_equal(node.submitblock(ToHex(block)), 'bad-cb-minerfund')
+       
+        # Using the Joshua scripts now works
+        for i in range(0, 12):
+            block = make_block_cb_post_numbers(JUDGES_SCRIPTS)
+            prepare_block(block)
+            assert_equal(node.submitblock(ToHex(block)), None)
+
         # Check replay protection is not enabled yet
         tx = CTransaction()
         tx.vin = [CTxIn(COutPoint(int(cointxid, 16), 1))]
-- 
2.39.5 (Apple Git-154)

